client/client-lib/src/api.rs

```diff
@@ -1,148 +1,615 @@
+///
+/// `params` will *not* be wrapped into a array, and must convert to one on its own.
+///
+/// ## Panics
+/// When `params` can't serialize to `serd_json::Value::Array` (which should be never)
+#[cfg(target_family = "wasm")]
+pub async fn with_erased_multi_params<Params, Res, E, F, Fut>(
+    params: &Params,
+    f: F,
+) -> result::Result<Res, E>
+where
+    Params: Serialize,
+    F: FnOnce(Vec<Value>) -> Fut,
+    Fut: Future<Output = result::Result<Res, E>>,
+{
+    let params_raw = match serde_json::to_value(params)
+        .expect("parameter serialization error - this should not happen")
+    {
+        // TODO: BUG: server side always expect an array Issue #1336
+        Value::Array(v) => Value::Array(v),
+        o => panic!("argument did not serialize to a json array: {o}"),
+    };
 
-    // TODO: find a better abstraction for all our API endpoints that allows different strategies and timeouts
-    /// Checks if there exists an offer for a payment hash
-    async fn offer_exists(&self, payment_hash: Sha256Hash) -> Result<bool>;
+    // TODO: BUG: server side always expect an array Issue #1336
+    f(vec![params_raw]).await
+}
 
-    /// Fetch the current consensus block height (trailing actual block height)
-    async fn fetch_consensus_block_height(&self) -> Result<u64>;
+/// Run `f` with  typed `params` converted to [`Vec<serde_json::Value>`].
+///
+/// `params` will *not* be wrapped into a array, and must convert to one on its own.
+///
+/// ## Panics
+/// When `params` can't serialize to `serd_json::Value::Array` (which should be never)
+#[cfg(not(target_family = "wasm"))]
+pub async fn with_erased_multi_params<Params, Res, E, F, Fut>(
+    params: &Params,
+    f: F,
+) -> result::Result<Res, E>
+where
+    Params: Serialize + Send + Sync,
+    F: FnOnce(Vec<Value>) -> Fut,
+    Fut: Future<Output = result::Result<Res, E>> + Send + Send,
+{
+    let params_raw = match serde_json::to_value(params)
+        .expect("parameter serialization error - this should not happen")
+    {
+        // TODO: BUG: server side always expect an array Issue #1336
+        Value::Array(v) => Value::Array(v),
+        o => panic!("argument did not serialize to a json array: {o}"),
+    };
 
-    /// Fetch the expected peg-out fees given a peg-out tx
-    async fn fetch_peg_out_fees(
+    // TODO: BUG: server side always expect an array Issue #1336
+    f(vec![params_raw]).await
+}
+
+#[cfg_attr(target_family = "wasm", async_trait(? Send))]
+#[cfg_attr(not(target_family = "wasm"), async_trait)]
+pub trait IFederationApiFed: IFederationApi {
+    async fn request_fed_with_strategy<MemberRet: serde::de::DeserializeOwned, FedRet: Debug>(
         &self,
-        address: &Address,
-        amount: &Amount,
-    ) -> Result<Option<PegOutFees>>;
+        mut strategy: impl QueryStrategy<MemberRet, FedRet> + Send,
+        method: String,
+        params: Vec<Value>,
+    ) -> FedResult<FedRet> {
```

Comment:
Hard to say. Its just an extension trait for `IFederationApi` ... so maybe it's OK? It will be a bit weird if they are named completely differently, I think.

Reply:
I'm ok with it. Might call it `FederationApiExt`, but not a blocker.

Reply:
if it's not object safe, I'd be in favor of `FederationApiExt`. It's still short and it's similar to `IFederationApi`

Reply:
wait, why isn't `IFederationApiFed` object safe? @elsirion

Reply:
@mxz42 `<MemberRet: serde::de::DeserializeOwned, FedRet: Debug>` I think

Reply:
Will do a rename to `FederationApiExt` in post-PR, along with other renames.